<a href="https://colab.research.google.com/github/sonupp13/MY-PROJECTS-IN-NATURAL-LANGUAGE-PROCESSING/blob/main/FAKE_NEWS_DETECTION_PROJECT_SONU.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

The FakeNewsDatabase dataset contains news in six different domains: technology, education, business, sports, politics, and entertainment. The legitimate news included in the dataset were collected from a variety of mainstream news websites predominantly in the US such as the ABCNews, CNN, USAToday, NewYorkTimes, FoxNews, Bloomberg, and CNET among others. The fake news included in this dataset consist of fake versions of the legitimate news in the dataset, written using Mechanical Turk. More details on the data collection are provided in section 3 of the paper.

The Celebrity dataset contain news about celebrities (actors, singers, socialites, and politicians). The legitimate news in the dataset were obtained from entertainment, fashion and style news sections in mainstream news websites and from entertainment magazines websites. The fake news were obtained from gossip websites such as Entertainment Weekly, People Magazine, RadarOnline, and other tabloid and entertainment-oriented publications. The news articles were collected in pairs, with one article being legitimate and the other fake (rumors and false reports). The articles were manually verified using gossip-checking sites such as "GossipCop.com", and also cross-referenced with information from other entertainment news sources on the web.

In [2]:
import csv

with open('/content/drive/MyDrive/Colab Notebooks/news.csv', 'r') as file:
    reader = csv.reader('/content/drive/MyDrive/Colab Notebooks/news.csv')
    data = []
    for row_number, row in enumerate(reader, start=1):
        if row_number != 1292&6035:
            data.append(row)


In [3]:
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/news.csv")
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
df.columns

Index(['Unnamed: 0', 'title', 'text', 'label'], dtype='object')

In [5]:
columns_to_drop = ['Unnamed: 0']
df = df.drop(columns=columns_to_drop)

In [6]:
df.inplace=True
df.head()

title  \
0                       You Can Smell Hillary’s Fear   
1  Watch The Exact Moment Paul Ryan Committed Pol...   
2        Kerry to go to Paris in gesture of sympathy   
3  Bernie supporters on Twitter erupt in anger ag...   
4   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [7]:
df.describe()

title  \
count                            6335   
unique                           6256   
top     OnPolitics | 's politics blog   
freq                                5   

                                                     text label  
count                                                6335  6335  
unique                                               6060     2  
top     Killing Obama administration rules, dismantlin...  REAL  
freq                                                   58  3171

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6335 entries, 0 to 6334
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   6335 non-null   object
 1   text    6335 non-null   object
 2   label   6335 non-null   object
dtypes: object(3)
memory usage: 148.6+ KB


In [9]:
df['label'].unique()

array(['FAKE', 'REAL'], dtype=object)

In [10]:
df['label'].value_counts()

REAL    3171
FAKE    3164
Name: label, dtype: int64

In [11]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df['label'] = le.fit_transform(df['label'])
df.head()

title  \
0                       You Can Smell Hillary’s Fear   
1  Watch The Exact Moment Paul Ryan Committed Pol...   
2        Kerry to go to Paris in gesture of sympathy   
3  Bernie supporters on Twitter erupt in anger ag...   
4   The Battle of New York: Why This Primary Matters   

                                                text  label  
0  Daniel Greenfield, a Shillman Journalism Fello...      0  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...      0  
2  U.S. Secretary of State John F. Kerry said Mon...      1  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...      0  
4  It's primary day in New York and front-runners...      1

In [12]:
df.isnull().sum()

title    0
text     0
label    0
dtype: int64

In [13]:
messages= df.copy()
messages.reset_index(inplace=True)

In [14]:
messages.head()

index                                              title  \
0      0                       You Can Smell Hillary’s Fear   
1      1  Watch The Exact Moment Paul Ryan Committed Pol...   
2      2        Kerry to go to Paris in gesture of sympathy   
3      3  Bernie supporters on Twitter erupt in anger ag...   
4      4   The Battle of New York: Why This Primary Matters   

                                                text  label  
0  Daniel Greenfield, a Shillman Journalism Fello...      0  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...      0  
2  U.S. Secretary of State John F. Kerry said Mon...      1  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...      0  
4  It's primary day in New York and front-runners...      1

In [ ]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

from nltk.stem.porter import PorterStemmer

ps = PorterStemmer()

corpus = []

for i in range(0,len(messages)):

 review = re.sub('[^a-zA-Z]', ' ', messages['title'][i])
 review = review.lower()
 review = review.split()
 review = [ps.stem(word) for word in review if not word in stopwords.words('english')]
 review = ' ' .join(review)
 corpus.append(review)
corpus

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer
cv = TfidfVectorizer()
x = cv.fit_transform(corpus).toarray()

In [17]:
y= messages['label']

In [18]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score ,confusion_matrix

In [19]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

In [20]:
model= MultinomialNB()
model.fit(x_train,y_train)

MultinomialNB()

In [21]:
y_pred=model.predict(x_test)

In [22]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")


Accuracy: 0.81


In [23]:
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test, y_pred)
cm

array([[468, 147],
       [ 94, 558]])

In [37]:
from sklearn.svm import SVC

In [38]:
model=SVC(kernel='linear',random_state=0)
model.fit(x_train,y_train)

SVC(kernel='linear', random_state=0)

In [39]:
accuracy_score(y_pred,y_test)

0.8097868981846882

WE GOT HIGHEST ACCURACY OF 81% WITH NAIVE BAYES ALGORITHM
